### PR TITLE
HPCC-28865 Refactor ESDL script operation property and error handling

### PR DESCRIPTION
**Pull request by timothyklemm**
- Encapsulate the logic for operation properties that may be either an XPath or text. Prefer XPath to text.
- Add an error reporting interface to all operations to hide the usage of common member data from internally and externally generated reports.

Signed-off-by: Tim Klemm <tim.klemm@lexisnexisrisk.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [x] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [x] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

File: esp/esdllib/esdl_script.cpp [hunk 1]
```diff
@@ -93,10 +93,228 @@ class OptionalCriticalBlock
...
+ *
+ * An implementation encapsulates warning, error, and exception reporting logic for operations.
+ * Operations and their helpers can focus on why a message is being generated without having to
+ * know how messages are expected to be generated.
+ *
+ * Operations should generally use either `recordWarning`, `recordError`, or `recordException`.
+ * Operations that choose to ignore user defined `trace` values in output may invoke the
+ * *untraceable* variants. Factory functions should use the *untraceable as* variants.
+ */
+interface IEsdlOperationMessenger
```
Comment (afishbeck) [inline esp/esdllib/esdl_script.cpp:108]: I would include Trace in the name Messenger could mean too many things.

File: esp/esdllib/esdl_script.cpp [hunk 2]
```diff
@@ -2797,6 +2974,54 @@ class CEsdlFunctionRegister : public CInterfaceOf<IEsdlFunctionRegister>
     }
 };
 
+class CEsdlOperationMessenger : public IEsdlOperationMessenger
```
Comment (afishbeck) [inline esp/esdllib/esdl_script.cpp:2977]: Can we name this CEsdlOperationDefaultTraceMessenger?  If we ever use it for anything else we can rename it then.

**Review by afishbeck — CHANGES_REQUESTED** [review]
@timothyklemm some questions / comments / suggestions.

File: esp/esdllib/esdl_script.cpp [hunk 3]
```diff
@@ -93,10 +93,163 @@ class OptionalCriticalBlock
...
+     * @param msg
+     */
+    virtual void recordException(int code, const char* msg) const = 0;
+
+};
+
+/**
+ * @brief Encapsulation of an operation property that may be given as either an XPath or literal.
+ *
+ * Properties may be required or optional. Failure setting a required property rccords an error
```
Comment (afishbeck) [inline esp/esdllib/esdl_script.cpp:141]: typo "records"

File: esp/esdllib/esdl_script.cpp [hunk 4]
```diff
@@ -93,10 +93,163 @@ class OptionalCriticalBlock
...
+    }
+
+    inline bool isXPath() const
+    {
+        return xpath;
+    }
+
+    inline bool isLiteral() const
+    {
+        return !literal.isEmpty();
```
Comment (afishbeck) [inline esp/esdllib/esdl_script.cpp:173]: Actually through the usage I've seen, if xpath is set then the literal value is not used.  Therefore functionally wouldn't isLiteral() be (!isXpath() && !literal.isEmpty())?

**Review by afishbeck — CHANGES_REQUESTED** [review]
@timothyklemm looks good.  really cleans up the code where XPATH or LITERAL is supported.  But there are two minor comments.

File: esp/esdllib/esdl_script.cpp (hunk 4 above)
Comment (timothyklemm) [reply esp/esdllib/esdl_script.cpp:173]: The set method enforces that `xpath` and `literal` cannot both be populated. Testing both is unnecessary unless there is concern of set changing to allow both values.

Comment (afishbeck) [reply esp/esdllib/esdl_script.cpp:173]: Ok, yes I missed that.

**Comment by timothyklemm** [comment]
@afishbeck I fixed the typo and squashed.

**Review by afishbeck — APPROVED** [review]
@timothyklemm looks good.

**Comment by timothyklemm** [comment]
@ghalliday please merge.

**Review by ghalliday — APPROVED** [review]

**Comment by ghalliday** [comment]
Looks good.  I will merge to 9.0.0 though rather than 8.12.x since it is refactoring rather than fixing an issue.  Will merge once other changes are upmerged.

**Pull request merged**